### ISSUE TITLE
[BUG] (short description)

### ISSUE THREAD

**Issue by Randwic** [bigtreetech/BIGTREETECH-TouchScreenFirmware#2295]
### Steps to reproduce

1. Try print with Sd TFT port
2. Try print with Isb TFT port

**Expected behavior**

**Actual behavior**

### Additional Information
* Hello,

first i'm sorry for my english but i'm french and again sorry if i'm in the wrong section but this is the first time i've used a request on github.

I recently installed on my FlSun super Racer a Skr v2 card and a Tft70 from Biqu. However since that time I have had bugs in my printing sometimes the printer stops with the M21 error or the M1 error without having configured a pause in the Slicer. It has also happened to me that printing stops without error on the TFT. I have tried printing with the SD card as well as the USB from the TFT but I have the same concerns however I have no problem securing the USB stick on the motherboard. I have the latest version of TFT as well as the latest version on the motherboard would you have any idea of ​​the problem thanks in advance.

**Comment by oldman4U** [comment]
This is most likely a SKR problem and not a problem of the TFT firmware. Please update the title to something meaningful and close the ticket once you do not need it anymore.

Thank you

**Comment by Guilouz** [comment]
@oldman4U Nope it's also happen with Artillery Ruby Motherboard.

No issue with same Gcode when printing from motherboard port or with Octoprint. And no issue with official release from April 10.
And M1/21 happens randomly when printing with the SD or USB port of the TFT.

It seems that is recent PR cause this but impossible to find it. 

This is what I have already tested without success : 
- Use shielded cable to connect the RS232 port of the TFT to the motherboard
- Only use one serial port P1 in config, all others are disabled
- Also try to disable Emulated M109 And M190 settings

**Comment by radek8** [comment]
If you are using a high-speed serial port, try reducing the port speed to 115200.
Lay the serial cable between the SKR and the TFT as far as possible from the stepper motors and the TMC controller.
Be sure to use a shielded serial cable!
This is interference that occurs on the serial cable, damaging existing communications. I used to have a similar problem.
TFT sent to Marlin code M105, but Marlin received M10 and M1.
Take a look at the serial line communication errors here:
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/1297

**Comment by Guilouz** [comment]
@radek8 But how do you explain that with the same hardware configuration but with the official TFT firmware in release section the problem does not appear?

And for M21 ?

And as already said I already use a shielded cable (on your advice).

**Comment by radek8** [comment]
I don't know. Is it a distorted M221 command?
In my case, the M105 was damaged in format
M10105, M1, M10N105, M1005, MM105 ...

Do you have the option to log communication and, in the event of an error, check which commands have been sent and received?
This would help a lot to find where the problem came from.

**Comment by Guilouz** [comment]
What can I check the communication with? Because this only happens when printing from SD or USD TFT.

**Comment by radek8** [comment]
My problem arose a year ago also after some update.
But the update did not have to damage the FW. Could it be enough that the frequency of sending the M105 command changed and this could have an effect on the error rate?
I don't know, it's just my estimates.
After replacing the shielded cable, the problem was solved. So the problem was not directly in the FW, even though it appeared after the update.

**Comment by radek8** [comment]
I connected WiFi to TFT.
Then I connected the printer as a virtual serial port to a PC and monitored the communication in Pronterface.
But this is how he will see what TFT is sending and what Marlin is answering.
He won't see what Marlin actually accepted. But maybe that would be enough.

You will need to compile New FW with enabled
#define DEBUG_SERIAL_COMM
and the set port number you will monitor (for example WIFI port 2).

**Comment by radek8** [comment]
It would be better to watch Marlin, but I don't know if Marlin sends serial communication to all ports, such as USB. Then it would be enough to connect via USB to Pronterface and start printing from TFT. He would like to try it.

**Comment by radek8** [comment]
Debug Communications must be enabled in Pronterface

**Comment by kisslorand** [comment]
I have this M21/M1 issue also. It appeared only after my SKR v1.4 Turbo died and switched to SKR v2. More than that, my BlTouch deployed randomly during print. I had none of that with SKR v1.4 Turbo.
It happened on a Artillery SWX1 printer. What I did is, I rearranged the cables/pins so the "servo" signal is now wired in the ribbon cable so it sits between a VCC and GND wire so it is shielded at some degree. No problems with random BlTouch deploy since then.
It is enough just to power on my printer and let it idle. At random times I have a popup telling me "SD init OK". It is the response from MArlin to M21. Beats me...

**Comment by Guilouz** [comment]
@kisslorand I have also SKR2 but some builds from master work perfectly, up to a certain PR that I can't identify.

**Comment by radek8** [comment]
The TFT sends M220 and M221 commands at regular intervals to update the status.
This command on M1 or M21 is most likely to be corrupted.
The strange thing is that SKR2 owners are complaining about this problem.

**Comment by Guilouz** [comment]
@radek8 Also Artillery Ruby owners on Sidewinder X2

**Comment by kisslorand** [comment]
What MCU is on this Ruby?

**Comment by Guilouz** [comment]
@kisslorand STM32F401RCT6

**Comment by kisslorand** [comment]
So it's the same family, 32F4xx.

**Comment by Guilouz** [comment]
I think it's not related to this because with latest Marlin bugfix and TFT official release firmware there is no issue, only with master.

**Comment by kisslorand** [comment]
> The TFT sends M220 and M221 commands at regular intervals to update the status. This command on M1 or M21 is most likely to be corrupted. The strange thing is that SKR2 owners are complaining about this problem.

What could make only these messages to get corrupted?
I was watching my printer with Pronterface, I saw no "Unknown command". How can messages to get corrupted only to good commands?

**Comment by Guilouz** [comment]
@kisslorand Yes I do not understand either, if there were corrupted orders it will be the case of several orders, while the prints are perfect, no temperature drops or whatever.

**Comment by radek8** [comment]
Can't it also matter how Marlin and the MCU use Buffer?
Can you try to sound some Buffer values and test if that doesn't solve the problem?

**Comment by Guilouz** [comment]
@radek8 This is config buffer I use : 

// The number of linear moves that can be in the planner at once.
// The value of BLOCK_BUFFER_SIZE must be a power of 2 (e.g., 8, 16, 32)
#if BOTH(SDSUPPORT, DIRECT_STEPPING)
  #define BLOCK_BUFFER_SIZE  8
#elif ENABLED(SDSUPPORT)
  #define BLOCK_BUFFER_SIZE 64
#else
  #define BLOCK_BUFFER_SIZE 32
#endif

// @section serial

// The ASCII buffer for serial input
#define MAX_CMD_SIZE 96
#define BUFSIZE 32

// Transmission to Host Buffer Size
// To save 386 bytes of PROGMEM (and TX_BUFFER_SIZE+3 bytes of RAM) set to 0.
// To buffer a simple "ok" you need 4 bytes.
// For ADVANCED_OK (M105) you need 32 bytes.
// For debug-echo: 128 bytes for the optimal speed.
// Other output doesn't need to be that speedy.
// :[0, 2, 4, 8, 16, 32, 64, 128, 256]
#define TX_BUFFER_SIZE 64

// Host Receive Buffer Size
// Without XON/XOFF flow control (see SERIAL_XON_XOFF below) 32 bytes should be enough.
// To use flow control, set this buffer size to at least 1024 bytes.
// :[0, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048]
//#define RX_BUFFER_SIZE 1024

**Comment by radek8** [comment]
You have higher values than my settings, which can reduce your free RAM. Is that so, or am I wrong?

My settings are this.

// The number of linear moves that can be in the planner at once.
// The value of BLOCK_BUFFER_SIZE must be a power of 2 (e.g., 8, 16, 32)
#if BOTH(SDSUPPORT, DIRECT_STEPPING)
  #define BLOCK_BUFFER_SIZE  16
#elif ENABLED(SDSUPPORT)
  #define BLOCK_BUFFER_SIZE 16
#else
  #define BLOCK_BUFFER_SIZE 16
#endif

// @section serial

// The ASCII buffer for serial input
#define MAX_CMD_SIZE 96
#define BUFSIZE 16

// Transmission to Host Buffer Size
// To save 386 bytes of PROGMEM (and TX_BUFFER_SIZE+3 bytes of RAM) set to 0.
// To buffer a simple "ok" you need 4 bytes.
// For ADVANCED_OK (M105) you need 32 bytes.
// For debug-echo: 128 bytes for the optimal speed.
// Other output doesn't need to be that speedy.
// :[0, 2, 4, 8, 16, 32, 64, 128, 256]
#define TX_BUFFER_SIZE 16

// Host Receive Buffer Size
// Without XON/XOFF flow control (see SERIAL_XON_XOFF below) 32 bytes should be enough.
// To use flow control, set this buffer size to at least 1024 bytes.
// :[0, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048]
//#define RX_BUFFER_SIZE 1024

#if RX_BUFFER_SIZE >= 1024
  // Enable to have the controller send XON/XOFF control characters to
  // the host to signal the RX buffer is becoming full.
  //#define SERIAL_XON_XOFF
#endif

**Comment by kisslorand** [comment]
0100 1101 0011 0001 0000 1010 - "M1\n"
0100 1101 0011 0010 0011 0001 0000 1010 - "M21\n"
0100 1101 0011 0010 0011 0010 0011 0000 0000 1010 - "M220\n"
0100 1101 0011 0010 0011 0010 0011 0001 0000 1010 - "M221\n"

These are the binary form of M1, M21, M220 and M221 commands. The difference is too much between M1/M21 and M220/M221.

I used the same buffer settings as I used for my SKR 1.4 Turbo with a MCU (LPC 1769) that has less of everything than 32F409 (SKR v2).

**Comment by Guilouz** [comment]
@radek8 With my settings VSCode report :

RAM:   [==        ]  20.4% (used 26728 bytes from 131072 bytes)
Flash: [==        ]  22.0% (used 230200 bytes from 1048576 bytes)

**Comment by kisslorand** [comment]
> @kisslorand Did you just remove " void " from Marlin?

Nope. There are several ways to do it.
- in M0_m1.cpp: 
```
void GcodeSuite::M0_M1() {
return;
...
```
- in gcode.cpp:
```
    case 'M': switch (parser.codenum) {

      #if HAS_RESUME_CONTINUE
        case 0:                                                   // M0: Unconditional stop - Wait for user button press on LCD
        case 1: /*M0_M1();*/ break;                                   // M1: Conditional stop - Wait for user button press on LCD
      #endif
```

**Comment by Guilouz** [comment]
@kisslorand I have disabled M0/M1 from Marlin, M1 command is now unknown from host (Pronterface) but always enabled when is sended from TFT Terminal.

I think it is also necessary to disable these functions on the TFT side as well.

**Comment by kisslorand** [comment]
I just tested it. When M1 is captured by the TFT (from gcode file or from terminal manual enter) the message is different than when you send M1 from Pronterface to Marlin. The messages I was getting are the ones that comes when Marlin receives M1.

In the meantime I discovered a bug, during printing I sent M1 from terminal, print paused and when I hit resume the nozzle went into the bed and print continued. I have a nice scratch now... 
The same happens if you have M1 in your gcode file.
So one thing is clear, M1 is not received by the TFT (not from gcode file, not from the motherboard).

**Comment by Guilouz** [comment]
Yes when M1 is sending by host we have a popup "M1 Stop" with choice to continue or to cancel. When it's sending from TFT Terminal, just a "pause" popup with one finish button and need to click to Resume button on printing screen to restart print.

Do you think if it's only removed from Marlin it will be good ?

**Comment by kisslorand** [comment]
I don't know about others but I had only the "M1 Stop" poupup, not the "pause" one.

**Comment by DCib** [comment]
I have the same M1/M0 commands pop up that pause my print or click the screen to continue. Mine also happen whether I'm printing or not. Doesn't that mean it cant be the buffer size in marlin and is something else ?

**Comment by africsnail** [comment]
Same issue here

**Comment by kisslorand** [comment]
Yes, it is most probably something else but for the moment there's no clue at all where to look for the source of the problem.

**Comment by kisslorand** [comment]
Doing some further investigation, it seems it's the TFT that sends M1 and M21 to Marlin. Marlin responds to these only on the serial port it receives it. So if I send it from Pronterface (serial by USB) I get response only on Pronterface.
Still looking for the cause of it...

**Comment by kisslorand** [comment]
I found a possible fix. For those who want to try replace line 70 in "SpeedControl.c" with:
```
speedQueryWait = (storeCmd("M220\n") == true) ? storeCmd("M221\n") : false;
````

**Comment by Guilouz** [comment]
I will try.

**Comment by kisslorand** [comment]
The idea is to send them separately not together.

**Comment by Guilouz** [comment]
It’s a confirmation that M1 is sent by TFT.
Here popup message with M0/M1 disabled in Marlin.
Command try to be send.

![2BDDAE5D-B766-437F-979C-9B7D51354658](https://user-images.githubusercontent.com/12702322/149006188-e3956819-aa6a-402c-b12b-6d4cdbed1f5f.jpeg)

**Comment by kisslorand** [comment]
Did this happen with the modifications suggested by me?

**Comment by Guilouz** [comment]
No not tried yet

**Comment by ACNC88** [comment]
I also have this problem after upgrading to SKR mini E3 V3.0.
I am using a TFT35 v3.0. The issue started after I swapped JUST the controller board, which makes me think the problem is with the board not the display. I have updated the firmware on both motherboard and display but the problem is still there (randomly).

**Comment by elgroso52** [comment]
Same problem for me after switching from an SKR miniE3 V1.2 to an SKR miniE3 V3.0. with a TFT35 V3 E3.

**Comment by digant73** [comment]
As suggested by kisslorand, if splitting M220 and M221 in two different transmissions solves the issue it is possible that the mainboard is not reading and buffering the whole "M220\nM221\n" in one shot. So after processing M220\n and sending back the reply to TFT, the TFT will also set infoHost.wait to "false" (once "ok" is received from mainboard) so the next gcode of the print file is sent to the mainboard while the mainboard will try to read M221\n. That could be the explanation of the issue.
With the exception of another multi-command in MachineParameter menu (MBL command), there are no other multi-command sent with storeCmd function (all the other multi-commands are sent with mustStoreScript function that will split the commands in multiple transmissions)

**Comment by kisslorand** [comment]
The problem is not only during print. If I turn on my printer and leave it on idle than at some point I will get a response from M21 (SD card OK). I even bought a second SKR 2, this one with the 429 MCU and the problem persists.

**Comment by digant73** [comment]
do you have the issue even splitting M220 and M221?

**Comment by kisslorand** [comment]
I haven't had the chance to try with M220 and M221 separated. My test rig is with MKS GEN_L and I haven't seen the issue there. 
I just swapped it out with SKR v2 and I want to check if the M0_M1 & M21 issue will be present on that test rig.

**Comment by kisslorand** [comment]
By enablin DEBUG in the TFT's FW I can see this:
```
>>M220
M221
<<wait
wait
<<FR:100%
FR:100%
<<ok P15 B15
<<echo:E0 Flow: 100%
echo:E0 Flow: 100%
<<ok P15 B15
<<wait
wait
```

With the changes I suggested I see this:
```
>>M220
<<FR:100%
FR:100%
<<ok P15 B15
>>M221
<<echo:E0 Flow: 100%
echo:E0 Flow: 100%
<<ok P15 B15
<<wait
wait
```

I really do not know if it makes a difference but it seems a cleaner communication, one command, a response, an "ok" than comes the next command.

**Comment by kisslorand** [comment]
After more than 12 hours with the M220 and M221 separated, there's no sign of M0_M1 nor M21.
I am keeping the test rig on and logging to a file all serial communications.

**Comment by V1EngineeringInc** [comment]
I'm seeing the same issue, and have been through several versions of SKR and TFT firmware on two printers. Many other users have also confirmed this in the V1Engineering Forums. 

Random M1, and it seems the newer firmware the Z keeps going, so resuming is no longer an option.

**Comment by kisslorand** [comment]
So, on my test rig after 24 hours there's still no sign of M0_M1 & M21 bug with M220 and M221 separated.
On the other hand I had a 2.5 hours print and while waiting for the bed to cool down M21 appeared on my printer which doesn't have the patch for M220 and M221 separation. (I do not know about M0_M1 because I have that removed from my printer's Marlin)

**Comment by V1EngineeringInc** [comment]
Removing M0 is not an option for us, we use it for tool changes.

**Comment by ACNC88** [comment]
I'm willing to try kisslorand's mod, but I'm still skeptical this has to do with the TFT. I used the same tft with 2 different minis v2.0 boards and never had this issue before. I only have the M1 message, and it does show up with the printer idle too. Another bizarre thing with this board is that, occasionally it pauses for a fraction of a second, then keeps going with the job. I'm not sure if it has to do with the linear advance, since I just started using it with this board. Im also going to compile my own version of the firmware, with changes to the buffer size as suggested by others here. This is quite annoying as I can't rely on my printer with this board in it and I don't plan on using the display in marlin mode.

**Comment by kisslorand** [comment]
>I'm still skeptical this has to do with the TFT

It might not come from the TFT but is already a fact that it comes from the serial where the TFT is connected to. Marlin responds (feedback) to M0, M1, M21 only to the port from where it came and it responds to the port where the TFT is connected. When the printer is idle, there is no other command sent to Marlin than M220 and M221. These are also sent during printing.
By separating M220 from M221 I just followed a common sense logic, there's nothing else. It's a theory that is yet to be proven right or wrong.

**Comment by ACNC88** [comment]
Yeah sorry, I agree with your reasoning. What I meant is, why the v2.0 has no issues handling M220 and 221 together? By the way can someone  from BIQU step in to help too?

**Comment by kisslorand** [comment]
More than 48 hours with my test rig on with M220 and M221 separated and not a single hiccup...

**Comment by mustang5269** [comment]
I made the changes and will test to see if it works.

**Comment by mustang5269** [comment]
After making changes, I've left mine on for 10 minutes and got the M1 stop message. I've also had a no sd card message pop up as well.

**Comment by BradisBowser** [comment]
I will make changes today and report back

**Comment by mustang5269** [comment]
I went back into marlin and set the buffers back to stock and I haven't had the errors pop up as of yet. I've noticed a few who raised the buffers as I did. I followed a skr mini e3 v3 guide who suggested raising them and I wonder if that's why people are having the same issue.

**Comment by mustang5269** [comment]
> if I'm not wrong, I had similar issues in the past with MKS GEN L using the default values:
> 
> 
> 
> #define BUFSIZE 4
> 
> #define TX_BUFFER_SIZE 0
> 
> 
> 
> and I solved that issue using the following settings in Configuration_adv.h:
> 
> 
> 
> #define BUFSIZE 16
> 
> #define TX_BUFFER_SIZE 16
> 
> 
> 
> With those values I never had any issue

I think this solved the issue for me as well.

**Comment by V1EngineeringInc** [comment]
We use the SKR Pro, and have not made any buffer changes. Still have the issue.

**Comment by kisslorand** [comment]
> After making changes, I've left mine on for 10 minutes and got the M1 stop message. I've also had a no sd card message pop up as well.

This is bad news. :(

How ironic is that after 60 hours of monitoring the serial communication with M220 and M221 separated I had not a single issue and now I left my test rig on for two hours with a FW that has not separated M220 and M221 and I had an M21 strike.

Regarding the buffers... I have BUFSIZE 16 and TX_BUFFER_SIZE  0 for two years now and never had any issue with M0, M1, M21 until I changed to SKR v2.

**Comment by V1EngineeringInc** [comment]
That seems at least a little promising. Some common thread.

After seeing the buffer post I started to dig deeper into the configuration_adv.h file and there are a ton of options that are fairly new. I would just not know where to start. Any sort of debugging we can do? 

https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/Configuration_adv.h#L2248

**Comment by kisslorand** [comment]
Well... I found something...
By accident a partial serial communication was logged on my PC. when M21 stroke. It can be seen that Marlin responds to M220 and than responds to M21. There's no response for M221.
Before and after the M21 appearance there is a response for M221 after the response to M220. For me it is clear that this pair of successive M220 + M221 is an issue. Having a partial log I cannot determine if TFT sent M221 or M21 but I am sure Marlin processed an M21 command and never had an M221 after that M220.

Here's a part of the log:
```
wait
wait
FR:100%
echo:E0 Flow: 100%
wait
wait
FR:100%
echo:No SD card
wait
FR:100%
echo:E0 Flow: 100%
wait
wait
```

I don't know if it is important but there's only one "wait" after "No SD card". In any other situation there's  two of them.

**Comment by V1EngineeringInc** [comment]
But when you were running separate commands, you still had the issue happen? 

I do get random sd card messages as well. I wonder what happens if we run off the USB stick instead.

**Comment by kisslorand** [comment]
> But when you were running separate commands, you still had the issue happen?

No, as I said earlier I had my test rig on for 60 hours with M220 and M221 separated, monitoring was permanently on and not had a single misbehave.

**Comment by V1EngineeringInc** [comment]
Oh, I can give it a shot. If that is the case, opening a PR request to have that changed either way is probably a good idea.

**Comment by mustang5269** [comment]
I haven't had any M1 errors pop up  but I do get a as card init error which I don't know why because they all work.

**Comment by kisslorand** [comment]
> opening a PR request to have that changed either way is probably a good idea.

I don't know what to say... There are some people here who are not happy if code is increased without a solid reason. The separation of M220 and M221 is not a certainty for the bug elimination as for @mustang5269 it didn't work. Who knows? Maybe together with a TX_BUFFER_SIZE increase it can make a difference for everybody. 
For me, on my test rig and on my printer there was not a single M21 as long as I had the M220 and M221 separated. I repeat, it is not a certainty yet.

**Comment by ACNC88** [comment]
I run off the USB drive and I still have the problem.
What files are you guys using to compile your own firmware, the V2.0 ones? There is no build files in the V3.0 folder, just precompiled FW.

**Comment by mustang5269** [comment]
> I run off the USB drive and I still have the problem.
> 
> What files are you guys using to compile your own firmware, the V2.0 ones? There is no build files in the V3.0 folder, just precompiled FW.

https://github.com/bigtreetech/Marlin/tree/SKR-mini-E3-V3.0-G0B1

**Comment by Guilouz** [comment]
> > opening a PR request to have that changed either way is probably a good idea.
> 
> I don't know what to say... There are some people here who are not happy if code is increased without a solid reason. The separation of M220 and M221 is not a certainty for the bug elimination as for @mustang5269 it didn't work. Who knows? Maybe together with a TX_BUFFER_SIZE increase it can make a difference for everybody. For me, on my test rig and on my printer there was not a single M21 as long as I had the M220 and M221 separated. I repeat, it is not a certainty yet.

I use #define TX_BUFFER_SIZE 64, so i think buffer is not the cause. Same configuration in Marlin and for a very long time without problems and the errors appeared suddenly following many PRs. Something has changed but what? That is the question...

**Comment by digant73** [comment]
@Guilouz Try to use the same value for:
#define BUFSIZE
#define TX_BUFFER_SIZE

E.g.
#define BUFSIZE 16
#define TX_BUFFER_SIZE 16

The speed on TFT was improved and that could be the reason of the issue in Marlin. The TFT fw has no bug.

**Comment by ACNC88** [comment]
@digant73 
I have just compiled the FW with buffers increase to 16 bits, printing as I write and will let you know how it goes....already printet 2 test cubes with no stops.

**Comment by kisslorand** [comment]
Guys, the issue not appearing after a few hours doesn't mean anything. I had my rig powered on for 60 hours without any problems, yet for @mustang5269 the patch didn't work.
Some of you might remember the bug when during print the flow rate went down to 10%. It was the same dang M221! Now it strikes back in another form.



> The speed on TFT was improved and that could be the reason of the issue in Marlin. 

I highly doubt that. I use my MKS TFT at this speed (72MHz) for over a year now and had no such issue and my test rig uses BTT TFT35 and I do not remember any speed improvements for that one.

**Comment by mustang5269** [comment]
I've left your fix and put the buffer values back to what BIGTREETECH had for stock and no issues so far. 🤷‍♂️

**Comment by kisslorand** [comment]
And what are those stock values?

**Comment by mustang5269** [comment]
![image](https://user-images.githubusercontent.com/16362892/149677134-e034d4f7-494e-443b-a70f-ea4b5862a981.jpeg)

**Comment by kisslorand** [comment]
So...
I had a look at Marlin regarding this disputed TX_BUFFER SIZE...
That setting in modern motherboards doesn't really do much, only if it set to high values (128, 256). The motherboards use their own, hardcoded settings.
For example BTT SKR mini E3 V3.0 uses 1024  for its serial TX buffer, regardless of what you set TX_BUFFER_SIZE.
SKR v2 doesn't account for TX_BUFFER_SIZE unless it is higher than 64 (128, 256).
Just compile your Marlin with different TX_BUFFER_SIZE and check how much extra RAM is used, you'll be surprised that to a certain value there's no increase.

**Comment by mustang5269** [comment]
> So...
> 
> I had a look at Marlin regarding this disputed TX_BUFFER SIZE...
> 
> That setting in modern motherboards doesn't really do much, only if it set to high values (128, 256). The motherboards use their own, hardcoded settings.
> 
> For example BTT SKR mini E3 V3.0 uses 1024  for its serial TX buffer, regardless of what you set TX_BUFFER_SIZE.
> 
> SKR v2 doesn't account for TX_BUFFER_SIZE unless it is higher than 64 (128, 256).
> 
> Just compile your Marlin with different TX_BUFFER_SIZE and check how much extra RAM is used, you'll be surprised that to a certain value there's no increase.

Good info. Ya got me then.

**Comment by ACNC88** [comment]
bumping the buffers to 16 bits did nothing for me, M1 error popped up after a 2h print was done with the machine idle.
I did kisslorand's changes to the TFT FW, printed a cal cube and t, hen let the printer sit overnight. No errors yet, now printing another 2h job. i will  let you know how that goes! PS: left the buffers at 16bit, since this should make any difference, right?

**Comment by ACNC88** [comment]
Still no errors after 6+h printing and perhaps 10h idling. I never went this far before. Maybe the end of a nightmare? Thanks @kisslorand.
Still it remains a mystery why the SKR mini V2.0 has no issues handling M220/M221 together....

![2CAA8EE6D7B7_1642447690291](https://user-images.githubusercontent.com/97620321/149827842-bdd2e601-c97d-40da-8b81-b2ee8e012bf9.jpg)

**Comment by BradisBowser** [comment]
> More than 48 hours with my test rig on with M220 and M221 separated and not a single hiccup... Anyone else (especially those who complained) cares to join testing the patch??? I am not doing it for myself... There will be no patch made into the FW on my behalf until there's no feedback from others.
> 
> > By the way can someone from BIQU step in to help too?
> 
> By the way, anyone who complained can step into and try the suggested patch ?
> 
> For those who are bothered by this bug and cannot make themself the necessary changes, just let me know what TFT you have and I'll compile a FW for you.

I've been struggling with this a bit. Any chance I could take you up on this offer? I'm running TFT35 E3 V3 with SKR mini E3 V3.

**Comment by kisslorand** [comment]
Sure. PR #2325 and PR #2291 are also included.

[BIGTREE_TFT35_V3.0_E3.27.x.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/7884573/BIGTREE_TFT35_V3.0_E3.27.x.zip)

**Comment by ACNC88** [comment]
So towards the end of my second print, the printer came to a halt, similar to what happens with the M1 stop, but without any dialog box coming up. Impossible to continue. I pushed the pause button but nothing happened. I'm totally fed up with this bug!

**Comment by BradisBowser** [comment]
> Sure. PR #2325 and PR #2291 are also included.
> 
> [BIGTREE_TFT35_V3.0_E3.27.x.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/7884573/BIGTREE_TFT35_V3.0_E3.27.x.zip) [config.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/7884583/config.zip) [language_en.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/7884585/language_en.zip)

Big thanks for all the work on this. New firmware is installed and starting an overnight print. Will report back any news.

**Comment by V1EngineeringInc** [comment]
Well I am about 12 hours in on two printers and so far no SD card notices and no M1 pauses. Normally I surely would have seen at least one by now.

 Seeing these new comments, I am very interested to see if they have errors overnight.

**Comment by BradisBowser** [comment]
Printer ran all night with no stops (during print or at idle). Starting another 11hr project now and feeling I'm optimistic.

**Comment by radek8** [comment]
@kisslorand 
Your edit seems to be working. Although a few users still report this error. For some, it solves the problem completely, for others it at least prolongs the interval of its occurrence. Most users will agree to merge this adjustment with the master branch.

Personally, I think the main problem is on the Marlin side, where incoming data is lost on the serial port. It can also be seen that most of the victims are on SKR-2
But it's just a guess.

**Comment by elgroso52** [comment]
I just tried with your new update via octoprint connected to the motherboard(SKR miniE3V3) , After 10 minutes "M1 STOP".
I now try the same printing via the USB port of the TFT 35 E3V3 screen...

**Comment by kisslorand** [comment]
@radek8 
I am into building/programming an arduino to monitor the serial wires between the TFT and mainboard,  a HW monitor. I will not rest until I have a definitive answer who's responsible, the TFT or the mainboard. I started to feel that my "patch" just avoids the bug, not eliminates it because I was able to avoid M1/M21 in other ways too.

**Comment by V1EngineeringInc** [comment]
I'm willing to test your theory's!

In 27 hours in on two printers with just the command separation fix, still good. a little more printing than idle, but they have not been power cycled.

**Comment by V1EngineeringInc** [comment]
A friend has a logger that he ran a few weeks back. All sorts of stuff was happening, here a capture that bothered us.

https://us2.dh-cdn.net/uploads/db5587/original/3X/d/9/d9124d7a20cf062aee7b0f228a5d9f7738eb8339.png

**Comment by kisslorand** [comment]
It seems the mainboard had hard times calculating the ARC. What MB is it?

**Comment by V1EngineeringInc** [comment]
SKR Pro, to be fair I don't remember that Gcode, but we run CNC's up to 8' so some of those segments can be very long. Could it just be travel time and/or full buffer? ( I know very little about how the board actually handles this communication).

**Comment by kisslorand** [comment]
Ah! It figures. G3 is the gcode for ARC segments, if they were very long than the RX buffer got full hence the "Busy..." message.

**Comment by V1EngineeringInc** [comment]
We figured as long as it gets a busy signal, it keeps the communication open and does not time out.

**Comment by kisslorand** [comment]
As you can see, new gcodes are sent only after an "ok" acknowledgement.

**Comment by kisslorand** [comment]
The hunt has begun:

![1](https://user-images.githubusercontent.com/50251547/150117583-90aaf79f-cb9a-4f74-88a4-b5c519ffd825.jpg)
![2](https://user-images.githubusercontent.com/50251547/150117604-d8eb133e-d865-433c-80ef-f03f2e5706de.jpg)

An Arduino Mega2560 is used as a HW sniffer. It directly monitors the RX and TX wires between the TFT and the motherboard.

**Comment by kisslorand** [comment]
### ...and BAMMMMMMMMMMM!
Just as I was writing the above comment M21 stroke and this time it is as clear as possible where's the problem.

---
---
TFT    >> M220
TFT    >> M221
Marlin >> FR:100%
Marlin >> ok P15 B15
Marlin >> echo:E0 Flow: 100%
Marlin >> ok P15 B15
Marlin >> wait
Marlin >> X:0.000 Y:0.000 Z:0.000 E:0.000 Count X:0 Y:0 Z:0
Marlin >>  T:-14.61 /0.00 B:-14.69 /0.00 @:0 B@:0
Marlin >> wait
TFT    >> M220
TFT    >> M221
Marlin >> FR:100%
Marlin >> ok P15 B15
Marlin >> **echo:SD card ok**
Marlin >> ok P15 B15
Marlin >> wait
Marlin >> X:0.000 Y:0.000 Z:0.000 E:0.000 Count X:0 Y:0 Z:0
TFT    >> M220
TFT    >> M221
Marlin >> FR:100%
Marlin >> ok P15 B15
Marlin >> echo:E0 Flow: 100%
Marlin >> ok P15 B15
Marlin >>  T:-15.00 /0.00 B:-15.00 /0.00 @:0 B@:0
Marlin >> wait
Marlin >> X:0.000 Y:0.000 Z:0.000 E:0.000 Count X:0 Y:0 Z:0
Marlin >> wait

---
---
It's obvious that the TFT sent M221 and Marlin processed it as M21 (because it never responded to M221 but it responded to an M21).
Now comes the hardest part, is it because the sending protocol is not respected by the TFT? Is it because of Marlin? Is it a HW design bug?
I don't know the answer but I guess it will not hurt to send M220 and M221 separately and respect the protocol by sending a new command **only** after receiving an acknowledgement ("ok") of the previous one.

**Comment by radek8** [comment]
I'll ask.
Have you tried to see if the error occurs at other serial port speeds?
So if 115200 had trouble trying to increase to 250,000.
Or vice versa, if the problem was ns 250000 reduce the speed to 115200.

**Comment by kisslorand** [comment]
No. The wires themself were monitored. This is HW monitoring what I've done.

**Comment by radek8** [comment]
Interesting article.
https://nuttx.yahoogroups.narkive.com/scNCwdg5/stm32f4-uart-dropouts-on-115200

**Comment by kisslorand** [comment]
Thx @radek, the article in the link you provided made me concentrate on DMA usage in Marlin for the MCU I have (STM32F4xx).
I found something interesting in Marlin:
> F4 supports one DMA for RX and another for TX, but Marlin will never do read and write at same time, so we use the same DMA for both.

Where F4 refers to the STM32F4xx family.

Here's what I think what happens:
 - After receiving the first command "M220", the MCU responds while the TFT keeps sending the second command "M221"
 - As stated above Marlin uses the same DMA for RX and TX, so while sending the response for M220 it will stop reading
 - If Marlin stops reading, sent bytes are lost and we end up with truncated messages received as M1 or M21

So it's not a bug in Marlin because this behaviour is an intended one, it's the TFT who doesn't respects it. 

In the end I would keep the M220 and M221 sending separated (respecting Marlin's protocol).

**Comment by kisslorand** [comment]
There's a commit (https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2291/commits/13f4e7ba57b881fbedb4e8b46d822699619884e0) regarding this M1/M21 issue.

**Comment by BradisBowser** [comment]
Awesome work on this one! First time I've been able to reliably print with TFT35 to SKR in touchscreen mode. Over 2 days of straight printing without a single hiccup. Big thanks again.

**Comment by V1EngineeringInc** [comment]
Same here, still no issues on either printer. I hope that gets merged soon! It is not showing up under the PR's though.

**Comment by kisslorand** [comment]
PR #2291

**Comment by V1EngineeringInc** [comment]
WOW that is a lot rolled up in one PR. I hope this goes through soon.

Thank you for all you have been doing!!!

**Comment by J0hanss0n** [comment]
That's why I asked for having separate PRs for unrelated issues ([Comment](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2291#issuecomment-1013918160)). Nevertheless the maintainers could consider cherry pick the corresponding commit.

**Comment by radek8** [comment]
Heretical question.
Custom commands and start G-code and end G-code are entered as a sequence of commands separated by a newline character.
Are they sent to Marlin at the same time, or each other? Can't a similar problem arise here?

**Comment by kisslorand** [comment]
Fortunately those mentioned commands are sent by a different procedure (`mustStoreScript()`). This procedure separates the commands and send them to the queue one by one.

**Comment by mralien** [comment]
@kisslorand thank you for great work!

One moment - I'm not sure if it related to splitting line 131 in coodinate.c or not (maybe it's just my local issue), but seems that in some cases sending empty line before M114 ends up in frozen printer. In my case it happens when I try to raise Z axis right after printer started. And after that nothing works (temps, fan, G28, etc.)

**Comment by kisslorand** [comment]
Thank you for your your observation. I do not know if it was the source of the problem but that line had a bug indeed. It's been fixed by commit https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2291/commits/78ae84ab9612f0fd8f890c9583b50dcd7fe2244c in PR #2291.

**Comment by ACNC88** [comment]
I'm glad this "patch" is working for SKR user but is there any feedback from SKR mini E3 V3.0 users? As I mentioned, with separated M220/M221 commands, I never got the M1 error but my printer froze in the middle of a job without the TFT reporting any error. The TFT was totally frozen as well and the only solution was to shut down the printer (and lose the job). This is a similar scenario reported by @Randwic "printing stops without error on the TFT".

As I mentioned before, it's hard to understand why the very same TFT (same firmware) can run the mini V2.0 with no issue whatsoever, whereas the mini V3.0 (same family of chipset) cannot handle it. To me the issue is most likely with the new board design or some sort of interference with the serial cable? In any case, I wanted to test whether changing the serial port baud rate would solve the problem. I have increased the BR to 250000, but that again resulted in the printer freezing shortly after the start of the job. 

I then reduced the BR to 57600 and the printed has been working flawless since 2 days. I left it idle for extended periods with no errors or freezes. This is using the ORIGINAL TFT firmware without separation of M221 and M220 commands. Despite the reduced BR  I have been printing at 60mm/s with no issues.  I am not sure whether going much faster is possible at this BR but I don't really need to at the present.

**Comment by kisslorand** [comment]
Why do you think your issue is because of Marlin interpreting commands wrong?

**Comment by kisslorand** [comment]
BTW, regardless of separating M220 and MM21 fixes or not the lost bytes on Marlin side, it is wrong to send more commands at once. Not using ADVANCED_OK on the TFT, it is unknown how many empty slots are left in the Marlin's buffer (defined by BUFFSIZE) . So, how do you know it is OK, there is room for more than one command? You don't! Hence it is a bug by itself.

**Comment by J0hanss0n** [comment]
@ACNC88 
Which slicer are you using? I had the freezing problem with SuperSlicer and GeniusPro printer. There's a bug using not simplified retraction paths leading to clogged cmd queues.

**Comment by ACNC88** [comment]
@kisslorand 
I am not questioning the validity of your patch or saying it should not implemented. All I'm saying I still had problems with it and running at lower BR fixes it for me. Perhaps it's my board, I don't know, it requires more investigation and I don't really have the time for this, I'd rather print stuff. For sure the V2.0 gave me none of this headaches, regardless of how many commands the TFT throws at it.

@J0hanss0n 
I'm using Cura, as I mentioned, I have never seen a printer freeze like that before. The other E3 (running same sliced gcode) with the mini is a reliable workhorse.

**Comment by kisslorand** [comment]
@ACNC88 
When the TFT is frozen, it means it is in an infinite loop (usually this happens when it is waiting for something that never happens and there's no guard against it). You can open a separate issue where with the help of others it might be fixed. 

There's a new patch that addresses also the M1/M221 issue. In case anyone wants to try it here it is: commit https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/2291/commits/6fd1853d9f88213a35cf8bd7d02742faddebf75f in PR #2291.

**Comment by V1EngineeringInc** [comment]
Unfortunately, I can't seem to work out how to test part of a PR. If I just get that entire PR is it in there? No offense, but honestly, I would much rather just test one issue at a time. I poked through the rest, and it seems pretty straightforward, but there are a lot of changes.

**Comment by kisslorand** [comment]
You can take the whole PR and compile it for yourself all the commits are included.

**Comment by kisslorand** [comment]
Prbuilt binaries are ready for PR #2291. Everyone can download theirs here: https://mega.nz/folder/Hx4mXKjY#WFvcTZWweq_VRW9pR7nr_g

**Comment by V1EngineeringInc** [comment]
Oddly enough, I can't heat up the extruder now, it also does not kick out an error about the temp not moving. I have not changed what is on the SKR Pro, and it worked fine earlier today. I compiled it myself and used the same config from the last test (known working).

**Comment by V1EngineeringInc** [comment]
I tried your prebuild bin's and neither heater starts?

**Comment by kisslorand** [comment]
A fix has been made, PR  #2291 is updated. It\s 1AM here, I will update the binaries in the morning.

**Comment by V1EngineeringInc** [comment]
So Far so good!

**Comment by V1EngineeringInc** [comment]
16 hours in still looks good. With everyone leaving on holiday for a few weeks, I doubt this will get merged any time soon. I plan on making a release using this as the base so we can actually use the TFT side....provided it gets through the day error free.

 A lot of time is spent telling people to just use the Marlin Side.


Thank you for making the TFT actually work!

**Comment by V1EngineeringInc** [comment]
What fix are you using/testing?

**Comment by elgroso52** [comment]
PR #2291

**Comment by Guilouz** [comment]
I'm going to test your PR, so far I don't have an M1 Stop error anymore because I disabled the functions in Marlin, but here I recompiled a Marlin build and I'm going to re-test.

**Comment by ACNC88** [comment]
@elgroso52 
E3V3.0 owner here, I also mentioned the fix didn't solve my problems with this board, maybe there is something else going on with this hardware.  My temporary solution was to reduce the BR to 57600. No issues since even though I'm running the unpatched TFT firmware.

**Comment by elgroso52** [comment]
@Guilouz
I tried disabling the features in Marlin. I no longer have "M1 stop" messages on the TFT screen but the printing stops anyway.
@ACN88
I have not tried this solution. I try as soon as I have time...

**Comment by Guilouz** [comment]
@elgroso52 Depending how you disable it, I have printed many files on my Delta @ 200mm/s without any error.

**Comment by mustang5269** [comment]
Since reflashing marlin with stock buffers and applying your fix by separating the two commands I haven't had the issue come back yet. I will keep testing.

**Comment by elgroso52** [comment]
@Guilouz 
Just try your solution in Marlin...

**Comment by V1EngineeringInc** [comment]
I think it would be helpful at this point to always note what screen you are using and what board is driving it. This is a lot of new information.

My SKR Pro's with V3 E3's is working with the 3.5.1. I will run the newer stuff in a few hours.

**Comment by kisslorand** [comment]
From what I see there are 3 cases here:
1. unwanted M21 on the motherboard and a message on the TFT
2. unwanted M1 on the motherboard and a message on the TFT
3. print stop (eventual TFT freeze) without any message on the TFT

I do not know if they are related but it would be nice to have in the reports the data of what error do you have with what motherboard. It might help...

@V1EngineeringInc beat me to it.... :)

**Comment by elgroso52** [comment]
In my case,
Before:
Ender 5, Marlin 2.0.9.3 with SK3 miniE3 V1.2 with TFT35 E3V3==> no problems
Right now:
Ender 5, Marlin 2.0.9.3 with SK3 miniE3 V3 with TFT35 E3V3==> just M1 stop errors with stop printing

**Comment by V1EngineeringInc** [comment]
Updated my post, 3.5.1 is your staged numbering system from you #2291 PR.

**Comment by Guilouz** [comment]
Just finished a print in 17h without issue.

SKR2 with TFT70

**Comment by kisslorand** [comment]
> @Guilouz Just try your solution in Marlin... and no positives results...

In this case you have a different problem, it's not the infamous M1/M21.

**Comment by elgroso52** [comment]
So I'm going to join the idea of ACNC88 who thinks the problem comes from the SK3 miniE3 V3...

**Comment by mustang5269** [comment]
Had an M1 message pop up today. The only thing I did was powere on and heat the nozzle so that I could unload some filament. Told it to cool down and message popped up. I agree something else is a problem.

**Comment by kisslorand** [comment]
I've also made hardware logs of the communication between the TFT and Marlin and I can confirm that Marlin is the one who loose bytes from messages. The TFT never sends unwanted M1 or M21. The patch of separating commands is just a try to avoid the bug on Marlin side. For some boards (SKR v2) it works.
Strange enough that on Marlin's repository no one complains abut this bug...

**Comment by auroraonline001** [comment]
SKR E3 V3 with TFT35 V3. Both recent upgrades to dead SKR E3 V1.2. Recent FW on both
Initially 'no SD card' mid print, yet print continues without issue, even with both SD (board and TFT) present, (printing from TFT)
Also getting these warning at idle.
no firmware changes, now getting M1 stop issues, luckily I have caught it and printing can continue.

I have a second Ender3, SKR E3 V2, TFT35 V3 (year older). Older FW on SKR, identical to above on TFT
No issues !

going to try @kisslorand bigfix tomorrow...

**Comment by kisslorand** [comment]
There's a new PR (#2365) that deals only with separating commands.
PR #2291 was broken apart to individual fixes and enhancements and it is no longer active.

**Comment by auroraonline001** [comment]
Used #2291 and all seems to be fixed. Run 1 print and left idle for a time. No SD messages. Or random pauses. 👍🏻

**Comment by auroraonline001** [comment]
> Used #2291 and all seems to be fixed. Run 1 print and left idle for a time. No SD messages. Or random pauses. 👍🏻

noticed since this update , control knob has lost LEDs. are the two related?

**Comment by V1EngineeringInc** [comment]
My leds are good, tft35 e3v3.

**Comment by adamfehnel** [comment]
I'm still having random M1 command issues, during prints and on idle.  I'm on Master for this firmware.  SKR mini e3 v3 with TFT35-E3 V3.0.1.  Am I missing something obvious?  It feels like this was resolved in February so I'm not sure what else could be the issue.

**Comment by Seelenkind** [comment]
> I'm still having random M1 command issues, during prints and on idle.  I'm on Master for this firmware.  SKR mini e3 v3 with TFT35-E3 V3.0.1.  Am I missing something obvious?  It feels like this was resolved in February so I'm not sure what else could be the issue.

The problem is not resolved.  It can still happen.  I deactivated the M0/M1 code in Marlin.

https://github.com/Seelenkind/BIGTREETECH-SKR-mini-E3

**Comment by radek8** [comment]
The same problem with the random pause is reported by users who print via octoprint or do not have TFT BTT.
In most cases, these are specific control boards with Marlin.
One of them is SKR mini E3 V3.

https://github.com/MarlinFirmware/Marlin/issues/21204
